src/converters/app-converter.ts

```diff
@@ -0,0 +1,388 @@
+ */
+export class AppConverter {
+    // Map of stack artifactId to StackConverter
+    public readonly stacks = new Map<string, StackConverter>();
+
+    public readonly manifestReader: AssemblyManifestReader;
+
+    constructor(readonly host: IStackComponent) {
+        this.manifestReader = AssemblyManifestReader.fromPath(host.assemblyDir);
+    }
+
+    convert() {
+        for (const stackManifest of this.manifestReader.stackManifests) {
+            const stackConverter = new StackConverter(this.host, stackManifest);
+            this.stacks.set(stackManifest.id, stackConverter);
+            this.convertStackManifest(stackManifest);
+        }
+    }
+
+    private convertStackManifest(artifact: StackManifest): void {
+        const dependencies = new Set<ArtifactConverter>();
+        for (const file of artifact.fileAssets) {
+            const converter = new FileAssetManifestConverter(this.host, file);
+            converter.convert();
+            dependencies.add(converter);
+        }
+
+        // TODO add docker asset converter
+        // for (const image of artifact.dockerAssets) {
+        // }
+
+        const stackConverter = this.stacks.get(artifact.id);
+        if (!stackConverter) {
+            throw new Error(`missing CDK Stack for artifact ${artifact.id}`);
+        }
+        stackConverter.convert(dependencies);
+    }
+}
+
+/**
+ * StackConverter converts all of the resources in a CDK stack to Pulumi resources
+ */
+export class StackConverter extends ArtifactConverter {
+    readonly parameters = new Map<string, any>();
+    readonly resources = new Map<string, Mapping<pulumi.Resource>>();
+    readonly constructs = new Map<ConstructInfo, pulumi.Resource>();
+    stackResource!: CdkConstruct;
+
+    constructor(host: IStackComponent, readonly stack: StackManifest) {
+        super(host);
+    }
+
+    public convert(dependencies: Set<ArtifactConverter>) {
+        const dependencyGraphNodes = GraphBuilder.build(this.stack);
+
+        // process parameters first because resources will reference them
+        for (const [logicalId, value] of Object.entries(this.stack.parameters ?? {})) {
+            this.mapParameter(logicalId, value.Type, value.Default);
+        }
+
+        for (const n of dependencyGraphNodes) {
+            if (n.construct.id === this.stack.id) {
+                this.stackResource = new CdkConstruct(`${this.host.name}/${n.construct.path}`, n.construct.id, {
+                    parent: this.host,
+                    // TODO: we could do better here. Currently the stack depends on all assets, but really
+                    // only individual resources should depend on individual assets
+                    dependsOn: this.stackDependsOn(dependencies),
+                });
+                this.constructs.set(n.construct, this.stackResource);
+                continue;
+            }
+
+            const parent = this.constructs.get(n.construct.parent!)!;
```

Comment:
Are those non-null assertions safe?

Reply:
Some legacy code, but I agree we should not do non-null assertions